source/ui/beamline/AMXYThetaControlMoveButton.cpp

```diff
@@ -0,0 +1,282 @@
+{
+	if(control == controlX_)
+		return;
+
+	if(contextMenu_) {
+		contextMenu_->deleteLater();
+		contextMenu_ = 0;
+	}
+
+	if(controlX_) {
+		disconnect(controlX_, 0, this, 0);
+	}
+
+	controlX_ = control;
+	if(controlX_) {
+		connect(controlX_, SIGNAL(destroyed()), this, SLOT(onControlXDestroyed()));
+		connect(controlX_, SIGNAL(connected(bool)), this, SLOT(checkStatus()));
+	}
+
+	setText(QString::number(currentStepSize()) % (controlX_ ? controlX_->units() : QString()));
+
+	// enabled / disabled:
+	checkStatus();
+}
+
+void AMXYThetaControlMoveButton::setControlY(AMControl *control2)
+{
+	if(control2 == controlY_)
+		return;
+
+	if(contextMenu_) {
+		contextMenu_->deleteLater();
+		contextMenu_ = 0;
+	}
+
+	if(controlY_) {
+		disconnect(controlY_, 0, this, 0);
+	}
+
+	controlY_ = control2;
+	if(controlX_) {
+		connect(controlY_, SIGNAL(destroyed()), this, SLOT(onControlYDestroyed()));
+		connect(controlY_, SIGNAL(connected(bool)), this, SLOT(checkStatus()));
+	}
+
+	setText(QString::number(currentStepSize()) % (controlX_ ? controlX_->units() : QString()));
+
+	// enabled / disabled:
+	checkStatus();
+}
+
+void AMXYThetaControlMoveButton::setControlTheta(AMControl *controlR)
+{
+	if(controlR == controlTheta_)
+		return;
+
+	if(contextMenu_) {
+		contextMenu_->deleteLater();
+		contextMenu_ = 0;
+	}
+
+	if(controlTheta_) {
+		disconnect(controlTheta_, 0, this, 0);
+	}
+
+	controlTheta_ = controlR;
+	if(controlX_) {
+		connect(controlTheta_, SIGNAL(destroyed()), this, SLOT(onControlThetaDestroyed()));
+		connect(controlTheta_, SIGNAL(connected(bool)), this, SLOT(checkStatus()));
+	}
+
+	setText(QString::number(currentStepSize()) % (controlX_ ? controlX_->units() : QString()));
+
+	// enabled / disabled:
+	checkStatus();
+}
+
+void AMXYThetaControlMoveButton::onButtonClicked()
+{
+	if(controlX_ && controlY_ && controlTheta_) {
+		controlX_->moveRelative(currentStepSize() * (directionXReversed_ ? -1.0 : 1.0) * sin((controlTheta_->value() + angleOffset_)/57.297));
+		controlY_->moveRelative(currentStepSize() * (directionYReversed_ ? -1.0 : 1.0) * cos((controlTheta_->value() + angleOffset_)/57.297));
+	}
+}
+
+
+void AMXYThetaControlMoveButton::onControlXDestroyed()
+{
+	setControlX(0);
+}
+void AMXYThetaControlMoveButton::onControlYDestroyed()
+{
+	setControlY(0);
+}
+void AMXYThetaControlMoveButton::onControlThetaDestroyed()
+{
+	setControlTheta(0);
+}
+
+void AMXYThetaControlMoveButton::checkStatus()
+{
+	if(controlX_ && controlX_->isConnected() && controlY_ && controlY_->isConnected() && controlTheta_ && controlTheta_->isConnected())
+		setEnabled(true);
+	else
+		setDisabled(true);
+
+}
+
+bool AMXYThetaControlMoveButton::setStepSizeIndex(int index)
+{
+	if(index<0 || index >= stepSizes_.count())
+		return false;
+	if(stepSizeIndex_ != index) {
+		emit stepSizeIndexChanged(stepSizeIndex_ = index);
+	}
+	setText(QString::number(currentStepSize()) % (controlX_ ? controlX_->units() : QString()));
```

Comment:
It is assumed that X and Y have the same units, or the geometry doesn't make sense.  That is noted in class description in the header file.